homeassistant/components/purpleair/config_flow.py

```diff
@@ -0,0 +1,136 @@
+"""Config flow for PurpleAir integration."""
+from __future__ import annotations
+
+from collections.abc import Mapping
+from typing import Any, cast
+
+from aiopurpleair import API
+from aiopurpleair.errors import InvalidApiKeyError, PurpleAirError
+import voluptuous as vol
+
+from homeassistant.const import CONF_API_KEY, CONF_LATITUDE, CONF_LONGITUDE, CONF_NAME
+from homeassistant.core import async_get_hass, callback
+from homeassistant.helpers import aiohttp_client, config_validation as cv
+from homeassistant.helpers.schema_config_entry_flow import (
+    SchemaCommonFlowHandler,
+    SchemaConfigFlowHandler,
+    SchemaFlowError,
+    SchemaFlowFormStep,
+)
+
+from .const import CONF_SENSOR_INDICES, DOMAIN, LOGGER
+
+CONF_API = "api"
+CONF_DISTANCE = "distance"
+
+DEFAULT_CONFIG_ENTRY_NAME = "PurpleAir"
+DEFAULT_DISTANCE = 5
+
+STEP_CHECK_API_KEY_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_KEY): cv.string,
+    }
+)
+
+
+@callback
+def async_get_step_by_coordinates_schema() -> vol.Schema:
+    """Get a schema for the "by_coordinates" step."""
+    hass = async_get_hass()
+    return vol.Schema(
+        {
+            vol.Inclusive(
+                CONF_LATITUDE, "coords", default=hass.config.latitude
+            ): cv.latitude,
+            vol.Inclusive(
+                CONF_LONGITUDE, "coords", default=hass.config.longitude
+            ): cv.longitude,
+            vol.Optional(CONF_DISTANCE, default=DEFAULT_DISTANCE): cv.positive_int,
+        }
+    )
+
+
+async def async_validate_by_coordinates(
+    handler: SchemaCommonFlowHandler, user_input: dict[str, Any]
+) -> dict[str, Any]:
+    """Validate input into the "by_coordinates" step."""
+    api = handler.flow_state[CONF_API]
+
+    try:
+        [nearest_sensor] = await api.sensors.async_get_nearby_sensors(
+            ["name"],
+            user_input[CONF_LATITUDE],
+            user_input[CONF_LONGITUDE],
+            user_input[CONF_DISTANCE],
+            limit_results=1,
+        )
+    except ValueError as err:
+        raise SchemaFlowError("no_sensors_near_coordinates") from err
+    except PurpleAirError as err:
+        LOGGER.error("PurpleAir error while getting nearby sensors: %s", err)
+        raise SchemaFlowError("unknown") from err
+    except Exception as err:  # pylint: disable=broad-except
+        LOGGER.exception("Unexpected exception while getting nearby sensors: %s", err)
+        raise SchemaFlowError("unknown") from err
+
+    return handler.options | {
+        CONF_NAME: nearest_sensor.name,
+        CONF_SENSOR_INDICES: [nearest_sensor.sensor_index],
```

Comment:
FYI, I'm storing sensor indices like this in anticipation of a future options flow that allows users to add/remove sensors to this same config entry (for batching API requests).